### ISSUE TITLE
 Error C4146: unary minus operation applied to unsigned type, result still unsigned

### ISSUE THREAD

**Issue by Dudly01** [pytorch/pytorch#29654]
## To Reproduce

1. Make libtorch 1.3.1 (CPU only) work with Visual Studio 2019 on Windows. (Set up PATH and project properties)
1. Run end-to-end example found [here](https://pytorch.org/cppdocs/frontend.html#end-to-end-example).
1. Get error C4146: "unary minus operation applied to unsigned type, result still unsigned".

The error traces back to:
libtorch\include\c10\util\llvmMathExtras.h
Line 395 `return -(UINT64_C(1)<<(N-1));` inside of:
```
 /// Gets the minimum value for a N-bit signed integer.
 inline int64_t minIntN(int64_t N) {
   assert(N > 0 && N <= 64 && "integer width out of range");

   return -(UINT64_C(1)<<(N-1));     
 }
```

 - PyTorch Version (e.g., 1.0): 
1.3.1 (downloaded zips on 12.11.2019)
 - OS (e.g., Linux):
 Windows
 - How you installed PyTorch (`conda`, `pip`, source): 
Extracted zip, set up dependencies inside MsVS2019
 - Any other relevant information:
The issue (and the possible solution) seems to be the same as in [this post](https://github.com/pytorch/pytorch/issues/15497).
So after deleting the '-' from the return value, it does no throw the error.

cc @ezyang @gchanan @zou3519 @jerryzh168 @peterjc123

**Comment by mruberry** [comment]
Sorry are you getting an error that prevents compilation or are you getting a warning during compilation, but the compilation completes successfully? 

What happens if you lower the warning level to 0?

**Comment by Dudly01** [comment]
It prevents compilation. If i were to run a much simpler code (see below) I would get the same error.

Simpler Code:
```
#include <torch/torch.h>
#include <iostream>

int main() {
  torch::Tensor tensor = torch::eye(3);
  std::cout << tensor << std::endl;
}
```

Visual Studio "output":
![kép](https://user-images.githubusercontent.com/24471678/68699648-d6c04180-0583-11ea-9b42-99e2b16d835d.png)

Upon deleting that minus sign before the return value (libtorch\include\c10\util\llvmMathExtras.h
Line 395), the simpler code works as expected.

(I am lowering the warning level to 0 now, as I have to look into how to do that)

**Comment by Dudly01** [comment]
Upon turning the warning level to zero it still can not compile, but ithe number of warnings (exclamation mark in yellow triangle on my uploaded photo) decreases to 40.

**Comment by mruberry** [comment]
Thanks for clarifying ;)

**Comment by Dudly01** [comment]
Lastly it happens both in release and debugging modes.

**Comment by peterjc123** [comment]
Similar with https://github.com/pytorch/pytorch/issues/15497. I'll fix this one.

**Comment by peterjc123** [comment]
Well, looks like that we fixed this one in https://github.com/pytorch/pytorch/pull/27596, but this fix is not merged into the v1.3.1 branch. You'll have to apply that patch manually.
